### PR TITLE
SAA-644 - Add migration endpoint

### PR DESCRIPTION
**Pull request by sioldham**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+          prisonId = request.prisonCode!!,
+        ),
+      ),
+    )
+
+    val response = webTestClient.migrateAppointment(request)!!
+    verifyAppointment(response)
+    verifyDatabase()
+
+    verify(eventsPublisher, times(1)).send(eventCaptor.capture())
```
Comment (sioldham) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:61]: Is this right? Do we want it publishing events?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/MigrateAppointmentController.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  private val appointmentService: AppointmentService,
+) {
+  @ResponseStatus(HttpStatus.CREATED)
+  @PostMapping()
+  @Operation(
+    summary = "Create an appointment or series of appointment occurrences",
+    description =
+    """
+    Create an appointment or series of appointment occurrences and allocate the supplied prisoner or prisoners to them.
+    Does not require any specific roles
```
Comment (sioldham) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/MigrateAppointmentController.kt:35]: Is the role actually going to be called XXX or is that a placeholder for TBC?

Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/MigrateAppointmentController.kt:35]: It's a placeholder. Intention was to get that role decided and added before you started the ticket but it'll happen after now in another ticket. For now, it's OK that there's no role restrictions on this

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt (hunk 1 above)
Comment (iterative-it) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:61]: Not for migrations, no. The `appointments.appointment-instance.created` event should be suppressed for migrated appointments so this should be changed to verify that no events of any sort have been published

Comment (sioldham) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+          prisonId = request.prisonCode!!,
+        ),
+      ),
+    )
+
+    val response = webTestClient.migrateAppointment(request)!!
+    verifyAppointment(response)
+    verifyDatabase()
+
+    verify(eventsPublisher, times(0)).send(eventCaptor.capture())
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:60]: You can use verifyNoInteractions instead here e.g. `verifyNoInteractions(eventsPublisher)`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  private fun verifyAppointment(response: Appointment) {
+    assertThat(response.id).isNotNull
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:64]: Can avoid a little bit of repetition here perhaps ...
```
with(response) {
  assertThat(id).isNotNull()
  assertThat(createdBy).isEqualTo("test-client")
  ....
}
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          prisonerNumber = request.prisonerNumber,
+          bookingId = 1,
+          prisonId = request.prisonCode!!,
+        ),
+      ),
+    )
+
+    val response = webTestClient.migrateAppointment(request)!!
+    verifyAppointment(response)
+    verifyDatabase()
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:58]: I'd suggest if you are going to check the database post request it is probably a good idea to do a pre-post assertion as well e.g. the db should be empty ???

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt [hunk 6]
```diff
@@ -115,18 +136,18 @@ class AppointmentService(
...
+      failIfPrisonCodeNotInUserCaseLoad(prisonCode!!)
+      failIfCategoryNotFound(categoryCode!!)
+      failIfLocationNotFound(inCell, prisonCode, internalLocationId)
+      failIfMissingPrisoners(prisonerNumbers, prisonerBookings)
+    }
 
     return AppointmentEntity(
-      categoryCode = categoryCode,
-      prisonCode = prisonCode,
+      categoryCode = categoryCode ?: "",
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt:149]: I might be missing something here but looking at the request these are mandatory (and also mandatory on the entity).  

Why would they be assigned as empty strings?

File: src/main/resources/migrations/common/V2023.05.03.2__add_is_migrated_column_to_appointment.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE appointment
```
Comment (matt-crowson-moj) [inline src/main/resources/migrations/common/V2023.05.03.2__add_is_migrated_column_to_appointment.sql:1]: Not essential.

Just a thought but we could probably change this to the current date in the file name given it is now the 9th???

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt (hunk 6 above)
Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt:149]: Actually this is a fair point. The prisonCode and internalLocationId properties can be made non null on the request object. They will always be supplied by the migration process as the latter is only nullable due to the in cell concept. Something NOMIS's data doesn't support. internalLocationId can also have a NotNull annotation. That should remove the need for this coalesce

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt (hunk 5 above)
Comment (sioldham) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:58]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt (hunk 3 above)
Comment (sioldham) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateAppointmentIntegrationTest.kt:60]: Done

File: src/main/resources/migrations/common/V2023.05.03.2__add_is_migrated_column_to_appointment.sql (hunk 7 above)
Comment (sioldham) [reply src/main/resources/migrations/common/V2023.05.03.2__add_is_migrated_column_to_appointment.sql:1]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt [hunk 8]
```diff
@@ -94,8 +115,8 @@ class AppointmentService(
     }
   }
 
-  private fun failIfMissingPrisoners(prisonerNumbers: List<String>, prisonerMap: Map<String, Prisoner>) {
-    prisonerNumbers.filter { number -> !prisonerMap.containsKey(number) }.let {
+  private fun failIfMissingPrisoners(prisonerNumbers: List<String>, prisonerBookings: Map<String, String?>) {
+    prisonerNumbers.filter { number -> !prisonerBookings.containsKey(number) }.let {
```
[the remark this answers is not in the document]
Comment (sioldham) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt:119]: Done

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**